File: src/core/alignment/anchors.cpp
#include "alignment/anchors.h"

#include "alignment/fuzz_ratio.h"
#include "alignment/normalizer.h"
#include "alignment/python_round.h"

#include <algorithm>
#include <cstddef>
#include <vector>

namespace subcue {
namespace {

// 复刻 core/anchors.py::_monotonic_subset：按行号与位置递增选择权重最大的锚点子集。
[[nodiscard]] QVector<Anchor> monotonicSubset(const QVector<Anchor> &anchors)
{
    if (anchors.isEmpty()) {
        return {};
    }
    const qsizetype count = anchors.size();
    QVector<double> bestScore(count);
    QVector<int> previous(count, -1);
    for (qsizetype current = 0; current < count; ++current) {
        const double weight = static_cast<double>(anchors[current].end - anchors[current].start + 1)
            + anchors[current].confidence;
        bestScore[current] = weight;
        for (qsizetype prior = 0; prior < current; ++prior) {
            if (anchors[prior].lineIndex < anchors[current].lineIndex
                && anchors[prior].end < anchors[current].start
                && bestScore[prior] + weight > bestScore[current]) {
                bestScore[current] = bestScore[prior] + weight;
                previous[current] = static_cast<int>(prior);
            }
        }
    }
    qsizetype index = 0;
    for (qsizetype i = 1; i < count; ++i) {
        if (bestScore[i] > bestScore[index]) {
            index = i;
        }
    }
    QVector<Anchor> selected;
    while (index >= 0) {
        selected.append(anchors[index]);
        index = previous[index];
    }
    std::reverse(selected.begin(), selected.end());
    return selected;
}

} // namespace

QVector<Anchor> AnchorFinder::findAnchors(const QStringList &lines, const std::u32string &stream)
{
    QVector<Anchor> candidates;
    const qsizetype lineCount = lines.size();
    const std::size_t streamSize = stream.size();
    for (qsizetype lineIndex = 0; lineIndex < lineCount; ++lineIndex) {
        const std::u32string target = Normalizer::normalizeCodepoints(lines.at(lineIndex));
        if (target.size() < 3) {
            continue;
        }

        std::vector<std::size_t> positions;
        for (std::size_t pos = stream.find(target); pos != std::u32string::npos;
             pos = stream.find(target, pos + 1)) {
            positions.push_back(pos);
        }
        if (positions.size() == 1) {
            candidates.append(Anchor{static_cast<int>(lineIndex), positions[0],
                positions[0] + target.size() - 1, 1.0});
            continue;
        }
        if (!positions.empty() || streamSize > 80'000) {
            continue;
        }

        const std::size_t lineDivisor = static_cast<std::size_t>(std::max<qsizetype>(1, lineCount));
        const std::size_t expected = static_cast<std::size_t>(pythonRound(
            static_cast<double>(lineIndex) / static_cast<double>(lineDivisor)
            * static_cast<double>(streamSize)));
        const std::size_t radius = std::max<std::size_t>(200, streamSize / lineDivisor * 3);
        const std::size_t left = expected > radius ? expected - radius : 0;
        const std::size_t right = std::min(streamSize, expected + radius);

        struct ScoredPosition {
            double similarity;
            std::size_t position;
        };
        std::vector<ScoredPosition> scored;
        const std::size_t lastPosition = right >= target.size() ? right - target.size() + 1 : 0;
        const std::size_t scanEnd = std::max(left, lastPosition);
        for (std::size_t pos = left; pos < scanEnd; ++pos) {
            scored.push_back({FuzzRatio::ratio(
                target.data(), target.size(), stream.data() + pos, target.size()) / 100.0, pos});
        }
        std::sort(scored.begin(), scored.end(),
            [](const ScoredPosition &a, const ScoredPosition &b) noexcept {
                if (a.similarity != b.similarity) {
                    return a.similarity > b.similarity;
                }
                return a.position > b.position;
            });
        if (!scored.empty() && scored[0].similarity >= 0.92
            && (scored.size() == 1 || scored[0].similarity - scored[1].similarity >= 0.04)) {
            candidates.append(Anchor{static_cast<int>(lineIndex), scored[0].position,
                scored[0].position + target.size() - 1, scored[0].similarity});
        }
    }
    return monotonicSubset(candidates);
}

} // namespace subcue

File: tools/python_ref/core/anchors.py
from __future__ import annotations

from dataclasses import dataclass

from rapidfuzz import fuzz

from core.normalizer import normalize_text


@dataclass(frozen=True, slots=True)
class Anchor:
    line_index: int
    start: int
    end: int
    confidence: float


def find_anchors(lines: list[str], stream: str) -> list[Anchor]:
    candidates: list[Anchor] = []
    for line_index, original in enumerate(lines):
        target = normalize_text(original)
        if len(target) < 3:
            continue
        positions: list[int] = []
        start = stream.find(target)
        while start >= 0:
            positions.append(start)
            start = stream.find(target, start + 1)
        if len(positions) == 1:
            candidates.append(Anchor(line_index, positions[0], positions[0] + len(target) - 1, 1.0))
            continue
        if positions or len(stream) > 80_000:
            continue
        expected = round(line_index / max(1, len(lines)) * len(stream))
        radius = max(200, len(stream) // max(1, len(lines)) * 3)
        left, right = max(0, expected - radius), min(len(stream), expected + radius)
        scored = [(fuzz.ratio(target, stream[pos : pos + len(target)]) / 100, pos) for pos in range(left, max(left, right - len(target) + 1))]
        scored.sort(reverse=True)
        if scored and scored[0][0] >= 0.92 and (len(scored) == 1 or scored[0][0] - scored[1][0] >= 0.04):
            candidates.append(Anchor(line_index, scored[0][1], scored[0][1] + len(target) - 1, scored[0][0]))
    return _monotonic_subset(candidates)


def _monotonic_subset(anchors: list[Anchor]) -> list[Anchor]:
    if not anchors:
        return []
    best_score = [len(str(anchor.end - anchor.start + 1)) + anchor.confidence for anchor in anchors]
    previous = [-1] * len(anchors)
    for current, anchor in enumerate(anchors):
        weight = anchor.end - anchor.start + 1 + anchor.confidence
        best_score[current] = weight
        for prior in range(current):
            if anchors[prior].line_index < anchor.line_index and anchors[prior].end < anchor.start and best_score[prior] + weight > best_score[current]:
                best_score[current] = best_score[prior] + weight
                previous[current] = prior
    index = max(range(len(anchors)), key=best_score.__getitem__)
    selected: list[Anchor] = []
    while index >= 0:
        selected.append(anchors[index])
        index = previous[index]
    return list(reversed(selected))
